File: Python/Problems/Problems/problem2405.py
import heapq
import math
import sys

import numpy as np

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem2405(Problem):
    def test(self) -> bool:
        s = "abacaba"

        expected = 4

        result = self.partitionString(s)

        print(result)

        return result == expected

    def partitionString(self, s: str) -> int:
        n = len(s)
        present = [False] * 26
        result = 1

        for i in range(n):
            id = ord(s[i]) - ord('a')
            if present[id]:
                result += 1
                present = [False] * 26
            present[id] = True

        return result

File: C++/Problems/Problems/problem_2405.cpp
#include "problem_2405.h"

bool problem_2405::test() {
    std::string s = "abacaba";

    int expected = 4;

    auto result = partitionString(s);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_2405::partitionString(std::string s) {
    int n = s.size();
    std::vector present(26, false);
    int result = 1;

    for (int i = 0; i < n; i++)
    {
        int id = s[i] - 'a';
        if (present[id])
        {
            result++;
            present = std::vector(26, false);
        }
        present[id] = true;
    }

    return result;
}
